pkg/validatormanager/validatormanager.go

```diff
@@ -31,36 +42,238 @@ func AddPoAValidatorManagerContractToAllocations(
-		remoteAddress,
+		managerAddress,
 		nil,
-		"initialize((bytes32,bytes32,uint64,uint8),address)",
+		"initialize((bytes32,uint64,uint8),address)",
 		params,
-		initialOwner,
+		ownerAddress,
+	)
+}
+
+// constructs p-chain-validated (signed) subnet conversion warp
+// message, to be sent to the validators manager when
+// initializing validators set
+// the message specifies [subnetID] that is being converted
+// together with the validator's manager [managerBlockchainID],
+// [managerAddress], and the initial list of [validators]
+func PoaValidatorManagerGetPChainSubnetConversionWarpMessage(
+	network models.Network,
+	aggregatorLogger logging.Logger,
+	aggregatorLogLevel logging.Level,
+	aggregatorQuorumPercentage uint64,
+	subnetID ids.ID,
+	managerBlockchainID ids.ID,
+	managerAddress common.Address,
+	convertSubnetValidators []txs.ConvertSubnetValidator,
+) (*warp.Message, error) {
+	validators := []warpMessage.SubnetConversionValidatorData{}
+	for _, convertSubnetValidator := range convertSubnetValidators {
+		validators = append(validators, warpMessage.SubnetConversionValidatorData{
+			NodeID:       convertSubnetValidator.NodeID[:],
+			BLSPublicKey: convertSubnetValidator.Signer.PublicKey,
+			Weight:       convertSubnetValidator.Weight,
+		})
+	}
+	subnetConversionData := warpMessage.SubnetConversionData{
+		SubnetID:       subnetID,
+		ManagerChainID: managerBlockchainID,
+		ManagerAddress: managerAddress.Bytes(),
+		Validators:     validators,
+	}
+	subnetConversionID, err := warpMessage.SubnetConversionID(subnetConversionData)
+	if err != nil {
+		return nil, err
+	}
+	addressedCallPayload, err := warpMessage.NewSubnetConversion(subnetConversionID)
+	if err != nil {
+		return nil, err
+	}
+	subnetConversionAddressedCall, err := warpPayload.NewAddressedCall(
+		nil,
+		addressedCallPayload.Bytes(),
+	)
+	if err != nil {
+		return nil, err
+	}
+	subnetConversionUnsignedMessage, err := warp.NewUnsignedMessage(
+		network.ID,
+		avagoconstants.PlatformChainID, // p-chain sign
+		subnetConversionAddressedCall.Bytes(),
+	)
+	if err != nil {
+		return nil, err
+	}
+	signatureAggregator, err := interchain.NewSignatureAggregator(
+		network,
+		aggregatorLogger,
+		aggregatorLogLevel,
+		ids.Empty, // primary network validators sign
+		aggregatorQuorumPercentage,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return signatureAggregator.Sign(subnetConversionUnsignedMessage, nil)
+}
+
+// calls poa manager validators set init method,
+// passing to it the p-chain signed [subnetConversionSignedMessage]
+// so as to verify p-chain already proceesed the associated
+// ConvertSubnetTx
+func PoAValidatorManagerInitializeValidatorsSet(
+	rpcURL string,
+	managerAddress common.Address,
+	privateKey string,
+	subnetID ids.ID,
+	managerBlockchainID ids.ID,
+	convertSubnetValidators []txs.ConvertSubnetValidator,
+	subnetConversionSignedMessage *warp.Message,
+) (*types.Transaction, *types.Receipt, error) {
+	type InitialValidator struct {
```

Comment:
we should move types out of the func

Reply:
no. the types are only there to be able to be able to properly call TxToMethodWithWarpMessage, but
it is no part of the function call, prefer not to receive a struct but the sequence of params